src/main/java/com/hivemq/extensions/handler/PubackInterceptorHandler.java

```diff
@@ -0,0 +1,441 @@
+    public void write(
+            @NotNull final ChannelHandlerContext ctx, @NotNull final Object msg, @NotNull final ChannelPromise promise)
+            throws Exception {
+        if (!(msg instanceof PUBACK)) {
+            super.write(ctx, msg, promise);
+            return;
+        }
+        if (!handleOutboundPuback(ctx, (PUBACK) msg, promise)) {
+            super.write(ctx, msg, promise);
+        }
+
+    }
+
+    private boolean handleOutboundPuback(
+            final @NotNull ChannelHandlerContext ctx,
+            final @NotNull PUBACK puback,
+            final @NotNull ChannelPromise promise) {
+        final Channel channel = ctx.channel();
+        if (!channel.isActive()) {
+            return false;
+        }
+
+        final String clientId = channel.attr(ChannelAttributes.CLIENT_ID).get();
+        if (clientId == null) {
+            return false;
+        }
+
+        final ClientContextImpl clientContext = channel.attr(ChannelAttributes.PLUGIN_CLIENT_CONTEXT).get();
+        if (clientContext == null || clientContext.getPubackOutboundInterceptors().isEmpty()) {
+            return false;
+        }
+        final List<PubackOutboundInterceptor> pubackOutboundInterceptors =
+                clientContext.getPubackOutboundInterceptors();
+
+        final PubackOutboundOutputImpl output = new PubackOutboundOutputImpl(configurationService, asyncer, puback);
+        final PubackOutboundInputImpl
+                input = new PubackOutboundInputImpl(new PubackPacketImpl(puback), clientId, channel);
+        final SettableFuture<Void> interceptorFuture = SettableFuture.create();
+        final PubackOutboundInterceptorContext interceptorContext =
+                new PubackOutboundInterceptorContext(PubackOutboundInterceptorTask.class,
+                        clientId, input, output, interceptorFuture, pubackOutboundInterceptors.size());
+
+
+        for (final PubackOutboundInterceptor interceptor : pubackOutboundInterceptors) {
+
+            if (interceptorFuture.isDone()) {
+                break;
+            }
+
+            final HiveMQExtension plugin = hiveMQExtensions.getExtensionForClassloader(
+                    (IsolatedPluginClassloader) interceptor.getClass().getClassLoader());
+
+            //disabled extension would be null
+            if (plugin == null) {
+                interceptorContext.increment();
+                continue;
+            }
+            final PubackOutboundInterceptorTask interceptorTask =
+                    new PubackOutboundInterceptorTask(interceptor, interceptorFuture, plugin.getId());
+
+            pluginTaskExecutorService.handlePluginInOutTaskExecution(
+                    interceptorContext, input, output, interceptorTask);
+        }
+
+        final OutboundInterceptorFutureCallback callback =
+                new OutboundInterceptorFutureCallback(output, puback, ctx, promise);
+        Futures.addCallback(interceptorFuture, callback, ctx.executor());
```

Comment:
imho this callback and the interceptorFuture are unnecessary.
Why is the ctx.fireChannelRead not called instead of interceptorFuture.set(null) in the PubackInboundInterceptorContext?